lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -1866,6 +1880,61 @@ derivePublicKey ctx (ApiT wid) (ApiT role_) (ApiT ix) = do
         k <- liftHandler $ W.derivePublicKey @_ @s @k @n wrk wid role_ ix
         pure $ ApiVerificationKey (getRawKey k, role_)
 
+postAnyAddress
+    :: forall n. ApiCredentials
+    -> Handler AnyAddress
+postAnyAddress body = do
+    (addr, addrType) <-
+            case (body ^. #spending, body ^. #staking) of
+                (Just (Credential (Left (ApiPubKey bytes))), Nothing) ->
+                    pure ( unAddress $
+                           CA.paymentAddress discriminant (spendingFromKey bytes)
+                         , EnterpriseDelegating )
+                (Just (Credential (Right (ApiScript (ApiT script)))), Nothing) ->
+                    pure ( unAddress $
+                           CA.paymentAddress discriminant (spendingFromScript script)
+                         , EnterpriseDelegating )
+                (Nothing, Just (Credential (Left (ApiPubKey bytes)))) -> do
+                    let (Right stakeAddr) =
+                            CA.stakeAddress discriminant (stakingFromKey bytes)
+                    pure ( unAddress stakeAddr, RewardAccount )
+                (Nothing, Just (Credential (Right (ApiScript (ApiT script))))) -> do
+                    let (Right stakeAddr) =
+                            CA.stakeAddress discriminant (stakingFromScript script)
+                    pure ( unAddress stakeAddr, RewardAccount )
+                (Just (Credential (Left (ApiPubKey bytes1))), Just (Credential (Left (ApiPubKey bytes2)))) ->
+                    pure ( unAddress $
+                           CA.delegationAddress discriminant (spendingFromKey bytes1) (stakingFromKey bytes2)
+                         , EnterpriseDelegating )
+                (Just (Credential (Left (ApiPubKey bytes))), Just (Credential (Right (ApiScript (ApiT script))))) ->
+                    pure ( unAddress $
+                           CA.delegationAddress discriminant (spendingFromKey bytes) (stakingFromScript script)
+                         , EnterpriseDelegating )
+                (Just (Credential (Right (ApiScript (ApiT script)))), Just (Credential (Left (ApiPubKey bytes)))) ->
+                    pure ( unAddress $
+                           CA.delegationAddress discriminant (spendingFromScript script) (stakingFromKey bytes)
+                         , EnterpriseDelegating )
+                (Just (Credential (Right (ApiScript (ApiT script1)))), Just (Credential (Right (ApiScript (ApiT script2)))) )->
+                    pure ( unAddress $
+                           CA.delegationAddress discriminant (spendingFromScript script1) (stakingFromScript script2)
+                         , EnterpriseDelegating )
+                (Nothing, Nothing) -> fail "At least one credential is required"
```

Comment:
That's one big chunk :grimacing: 
My recommendation would be to capture this already in the API data-type itself. So, instead of:

```hs
data ApiCredentials = ApiCredentials
    { spending :: !(Maybe Credential)
    , staking :: !(Maybe Credential)
    } deriving (Eq, Generic, Show)
```

have something along the lines of:

```hs
data ApiAddressData 
    = AddrBase Credential Credential
    | AddrPtr Credential Ptr
    | AddrEnterprise Credential
    | AddrBootstrap Address
```

this would be quite more readable and also get rid of the "impossible" case.

Reply:
good idea, indeed it made things simpler -> https://github.com/input-output-hk/cardano-wallet/pull/2253/commits/1071e78b24b4974a03f4dee7b88c8743e5327775